eclipsesource/cdtcloud-deploymentserver
# feat: Startup scripts & Docker images

## Changes Proposed:


- Merge all Dockerfiles into one
- Proper images for the separate packages and services in Docker
- Properly adjusted docker-compose services
- Added Demo Docker Image
- Added Demo script in root package.json
- Added scripts to start the services and increase usability
- Various fixes

## Issues Addressed:


- Closes #82

## Tests Performed:


- Compiled and started every single docker image
- Started the demo natively
- Tested the StartUp script (./cdtcloud)


## How to Test the Changes:

- Requirements:
  - Running on Host-System
    - NodeJS Version 16.5 or higher
    - yarn
    - Postgresql DB server
  - Running in Docker: 
    - docker
    - docker-compose (Supporting Version 3.7 or higher)
- Start with Docker (Option 1)
  1. Run `./cdtcloud`
  2. Select the `Run docker tools` option
  3. Run the demo
  4. Also test every image by running the services separately
- Start Native on Host-System (Option 2)
  1. Run `./cdtcloud` (needs to be run as Admin on Windows)
  2. Select the `First installation` option
  3. Run the `Run demo` option
- Testing
  - Go to http://localhost:3000 for the theia browser implementation
  - Select View -> Cdtcloud Widget to open the extension
  - Create or open a file (An example project like [this](https://create.arduino.cc/projecthub/EvdS/led-dice-885cf1) can be used for testing purposes)
  - Attach a supported device (for example an Arduino Uno)
  - The widget list should now show a selectable board
  - Hit deploy to compile and upload the code onto the selected device
  - Go to http://localhost:3001 to view the Admin Dashboard

## Known Issues and TODO List:


- [x] Arduino-cli installation via script and startup command
- [x] Database isn't automatically initiated yet on the first run. Currently needs to be executed separately.
- [ ] Running the arduino-cli in docker on windows does not recognize attached devices due to the `privileged` option not being supported.
- [ ] When building theia for the browser, the electron app sometimes does not work anymore. The same problem occurs the other way around. 

## How to Test PRs
1) First fetch the Pull-Request from the main repo, replacing `xx` with the ID of the PR:\
`git fetch git@github.com:eclipsesource/cdtcloud-deploymentserver.git pull/xx/head:pr_xx`
2) Checkout the Pull-Request, again replacing `xx` with the PR-ID:\
`git checkout pr_xx`
3) Perform the necessary tests for the PR
